File: moderate/jolly_jumpers.rb
File.open(ARGV[0]).each_line do |line|
  n, *s = line.chomp.split.map(&:to_i)
  jolly = true
  if n > 1
    u = Array.new(n - 1, false)
    (1...n).each do |i|
      x = (s[i] - s[i - 1]).abs
      if x >= n || x == 0 || u[x - 1]
        jolly = false
        break
      end
      u[x - 1] = true
    end
  end
  puts jolly ? 'Jolly' : 'Not jolly'
end

File: moderate/jolly_jumpers.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    next if ( $line =~ m/^\s$/ );
    chomp($line);
    my @s = split( / /, $line );
    my $n = shift @s;
    my $j = 1;
    if ( $n > 1 ) {
        my %u;
        for ( my $i = 1 ; $i < $n ; $i++ ) {
            my $x = abs( @s[$i] - @s[ $i - 1 ] );
            if ( $x >= $n || $x == 0 || $u{ $x - 1 } ) {
                $j = 0;
                last;
            }
            $u{ $x - 1 } = 1;
        }
    }
    printf( "%s\n", $j ? "Jolly" : "Not jolly" );
}
close(INFILE);
